### DISCUSSION TITLE
`Page.annots()`:  ValueError: orphaned object: parent is None

### DISCUSSION THREAD

**Discussion by cbm755** [pymupdf/PyMuPDF#2484] [Q&A]
I'm not sure what the expected behaviour is, but I'm getting a ValueError from `.annots()`:

```
>> import fitz
>> fitz.version
Out[3]: ('1.22.3', '1.22.0', '20230510000001')

>> doc = fitz.open("DELETE_ME_tam_file_with_runtime_annot_errors.pdf")

>> list(doc[0].annots())
Out[8]: ---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
File /usr/lib/python3.11/site-packages/IPython/core/formatters.py:706, in PlainTextFormatter.__call__(self, obj)
    699 stream = StringIO()
    700 printer = pretty.RepresentationPrinter(stream, self.verbose,
    701     self.max_width, self.newline,
    702     max_seq_length=self.max_seq_length,
    703     singleton_pprinters=self.singleton_printers,
    704     type_pprinters=self.type_printers,
    705     deferred_pprinters=self.deferred_printers)
--> 706 printer.pretty(obj)
    707 printer.flush()
    708 return stream.getvalue()

File /usr/lib/python3.11/site-packages/IPython/lib/pretty.py:393, in RepresentationPrinter.pretty(self, obj)
    390 for cls in _get_mro(obj_class):
    391     if cls in self.type_pprinters:
    392         # printer registered in self.type_pprinters
--> 393         return self.type_pprinters[cls](obj, self, cycle)
    394     else:
    395         # deferred printer
    396         printer = self._in_deferred_types(cls)

File /usr/lib/python3.11/site-packages/IPython/lib/pretty.py:640, in _seq_pprinter_factory.<locals>.inner(obj, p, cycle)
    638         p.text(',')
    639         p.breakable()
--> 640     p.pretty(x)
    641 if len(obj) == 1 and isinstance(obj, tuple):
    642     # Special case for 1-item tuples.
    643     p.text(',')

File /usr/lib/python3.11/site-packages/IPython/lib/pretty.py:410, in RepresentationPrinter.pretty(self, obj)
    407                         return meth(obj, self, cycle)
    408                 if cls is not object \
    409                         and callable(cls.__dict__.get('__repr__')):
--> 410                     return _repr_pprint(obj, self, cycle)
    412     return _default_pprint(obj, self, cycle)
    413 finally:

File /usr/lib/python3.11/site-packages/IPython/lib/pretty.py:778, in _repr_pprint(obj, p, cycle)
    776 """A pprint that just redirects to the normal repr function."""
    777 # Find newlines and replace them with p.break_()
--> 778 output = repr(obj)
    779 lines = output.splitlines()
    780 with p.group():

File ~/.local/lib/python3.11/site-packages/fitz/fitz.py:8526, in Annot.__repr__(self)
   8525 def __repr__(self):
-> 8526     CheckParent(self)
   8527     return "'%s' annotation on %s" % (self.type[1], str(self.parent))

File ~/.local/lib/python3.11/site-packages/fitz/fitz.py:2975, in CheckParent(o)
   2973 def CheckParent(o: typing.Any):
   2974     if getattr(o, "parent", None) == None:
-> 2975         raise ValueError("orphaned object: parent is None")

ValueError: orphaned object: parent is None
```

- - - - 

I don't get the error if I send types:
```
>> list(doc[0].annots(types=[fitz.PDF_ANNOT_WIDGET]))
Out[36]: []

>> list(doc[0].annots(types=[fitz.PDF_ANNOT_POPUP]))
Out[37]: []

>> list(doc[0].annots(types=[fitz.PDF_ANNOT_LINK]))
Out[38]: []
```

I also don't get the error if reference a page directly:
```
>> p = doc[0]

>> list(p.annots())
Out[42]: 
['FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'Ink' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'Ink' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf,
 'FreeText' annotation on page 0 of DELETE_ME_tam_file_with_runtime_annot_errors.pdf]

```
I can't think why `p.annots()` would be different than `doc[0].annots()`...

**Comment by cbm755** [comment]
Sorry I cannot share this PDF file, but its not a totally healthy file:
```
$ pdfinfo DELETE_ME_tam_file_with_runtime_annot_errors.pdf 
Creator:         BaKoMa TeX 11.80 29518P1573/366481723
Producer:        Lahore University of Management Sciences (LUMS), Lahore, Pakistan
CreationDate:    Wed Jun  7 12:07:10 2023 PDT
Custom Metadata: no
Metadata Stream: no
Tagged:          no
UserProperties:  no
Suspects:        no
Form:            AcroForm
Syntax Error: Can't get Fields array<0a>
JavaScript:      no
Pages:           24
Encrypted:       no
Page size:       612 x 792 pts (letter)
Page rot:        0
File size:       1713255 bytes
Optimized:       no
PDF version:     1.4
```

**Comment by cbm755** [comment]
somewhat surprisingly:
```
In [10]: doc.is_repaired
Out[10]: False
```

The first few lines look like this:
```
%PDF-1.4^M
%¡³Å×^M
1 0 obj^M
[/PDF/Text/ImageB/ImageC]^M
endobj^M
2 0 obj^M
<</AcroForm 174 0 R /Pages 4 0 R /Type/Catalog>>^M
endobj^M
3 0 obj^M
<</CreationDate(D:20230607190710)/Creator(BaKoMa TeX 11.80 29518P1573/366481723)/Producer(Lahore University of Management Sciences \(LUMS\), Lahore, Pakistan)>>^M
endobj^M
4 0 obj^M
<</Count 24/Kids[ 5 0 R  6 0 R  7 0 R  8 0 R  9 0 R  10 0 R  11 0 R  12 0 R  13 0 R  14 0 R  15 0 R  16 0 R  17 0 R  18 0 R  19 0 R  20 0 R  21 0 R  22 0 R  23 0 R  24 0 R  25 0 R  26 0 R  27 0 R  28 0 R ]/MediaBox[ 0 0 612 792]/Type/Pages>>^M
endobj^M
5 0 obj^M
<</Annots[ 176 0 R  178 0 R  181 0 R  183 0 R  185 0 R  187 0 R  189 0 R  191 0 R  193 0 R  195 0 R  198 0 R  200 0 R  201 0 R  203 0 R  205 0 R ]/Contents 34 0 R /Parent 4 0 R /ProcSet 1 0 R /Resources 29 0 R /Type/Page>>^M
endobj^M
6 0 obj^M
<</Annots[ 207 0 R ]/Contents 46 0 R /Parent 4 0 R /ProcSet 1 0 R /Resources 36 0 R /Type/Page>>^M
endobj^M
7 0 obj^M
<</Annots[ 222 0 R ]/Contents 49 0 R /Parent 4 0 R /ProcSet 1 0 R /Resources 48 0 R /Type/Page>>^M
endobj^M
8 0 obj^M
<</Annots[ 224 0 R  2335 0 R ]/Contents 52 0 R /Parent 4 0 R /ProcSet 1 0 R /Resources 51 0 R /Type/Page>>^M
endobj^M
9 0 obj^M
```
And there are many more `<</Annots` but I'm nervous about posting content.

Let me know if you need any other info from the file...

**Comment by JorjMcKie** [comment]
That's a lot of posts. There is no bug, and I think I can address what seems surprising at first sight.
Please let me transfer this to "Discussions" to have more room for details.

**Comment by JorjMcKie** [comment]
1. Selecting annotation types, e.g. `list(doc[0].annots(types=[fitz.PDF_ANNOT_WIDGET]))`, delivers empty lists in all your cases, because you chose types that MuPDF does not count as annotations for the iterator `.annots()`. This iterator ingores links, widgets and popups.
    - links and widgets have their own iterators
    - popups don't count, because they only exist as sort of a child of a real annotation.

2. There is a big difference between `page.something()` and `doc[0].something()`:
    - The second syntax: (1) creates the page, (2) executes `something()` (3) removes the page. If anything goes wrong executing the **method** in (2), the page will be deleted at that point so that (3) runs into problems.
    - `Page.annots()` is an iterator, which I implemented internally using 2 steps: I first make a list of annotation xref numbers, and secondly iterate over that list, yielding each annotation via `Page.load_annot(xref)`. I chose that approach, because users tried to delete annotations from within the iterator which was leading to interpreter crashes. While changing an iterator from within its execution body is wrong in any case, bringing down the interpreter is still not justifiable. So in this case, making that xref list plays the role of `something()`. When trying to do `page.load_annot(xref)`, the page is gone already.

5. `doc.is_repaired`: Document creation does execute **_necessary_** checks to confirm that the file is usable - _roughly_  if you will. It cannot and **_should not_** uncover all sorts of errors hidden in some corner of its content. Even if that were possible, it would be a performance nightmare. What PDF has under the umbrella "annotations" is a plethora of wildly different things (links, true annots, fields, popups). The xrefs of all of them will be found in the page's dictionary key `/Annots` (a PDF array of xref numbers). MuPDF will for example **_not confirm at open time_** whether the xrefs in these arrays exist at all or whether they all are any of the valid types ... for all 100 thousand pages. Obviously, this would entail millions of unnecessary checks.

**Reply by cbm755** [reply]
Thanks for the detailed answers.  All makes sense except 2.: I afraid I still don't get it.  I don't see how the internal details matter: how can the existence of a temporary variable `p`  possibly change the results?

Its some kind of magic: when I look at "**A**" and "**B**" below, I assume the interpreter is free to change my code in A into B...  Some fundamental assumption about programming is broken in my brain for this not be be so!

### A
```python
>> doc = fitz.open("f.pdf")
>> p = doc[0]
>> type(p)
fitz.fitz.Page

>> list(p.annots())

['FreeText' annotation on page 0 of f.pdf,
...
 'Ink' annotation on page 0 of f.pdf,
 'Ink' annotation on page 0 of f.pdf,
...
 'FreeText' annotation on page 0 of f.pdf]

>> del p
>> del doc
```

### B
```python
>> doc = fitz.open("f.pdf")
>> type(doc[0])
fitz.fitz.Page

>> list(doc[0].annots())

ValueError: orphaned object: parent is None
```

**Reply by cbm755** [reply]
I also considered:

## C ##
```
gen = p.annots()
list(gen)
# fine
```

## D ##
```
gen = doc[0].annots()
list(gen)
# boom!
```

**Reply by JorjMcKie** [reply]
Let me make another try to explain what happens. Here is a session that may help:
```python
In [1]: import fitz
In [2]: doc = fitz.open("new-annots-0.pdf")
In [3]: annots = []  # list of annots on page 0
In [4]: for a in doc[0].annots():  # this is unproblematic
   ...:     annots.append(a)
   ...:
In [5]: # This works!
In [6]: # But the list is mostly unusable, because the page is deleted now.
In [7]: # None of its items has a parent:
In [8]: len(annots)  # no problem, as we do not touch items here
Out[8]: 15

In [9]: print(annots)  # boom!
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
ValueError: orphaned object: parent is None
In [10]: tuple(annots)  # boom!
Out[10]: ---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
ValueError: orphaned object: parent is None

In [11]: annots[0]  # boom!
Out[11]: ---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
...
ValueError: orphaned object: parent is None
```

So it is **_not_** `doc[0].annots()` that throws `ValueError`! It is the illegal use of its yielded items after it has ended.

**Comment by cbm755** [comment]
I somewhat hesitate to muddy the waters further but the original error reported by one of my users is in fact with the `p`-form.  This is the code we run:
```python
    msgs = []
    # Any of these might indicate something more complicated than a scan
    # and hence we should be safe and just render the page.  We only try to
    # extract the bitmap under very conservative circumstances.  It is not
    # safe to assume that if there is a single image on the current page
    # then that is the scan - e.g., student annotates pdf using xournalpp and
    # then stamps a smiley-face `.png` there."
    if p.get_links():
        msgs.append("Has links")
    if list(p.annots()):
        msgs.append("Has annotations")
    if list(p.widgets()):
        msgs.append("Has fillable forms")
    if p.get_text("text"):
        msgs.append("Has text")
```
(here `p` is a `fitz.Page`).  The error was reported to me as:

```
  File "/home/ndntam/.local/lib/python3.10/site-packages/plom/scan/scansToImages.py", line 271, in try_to_extract_image
    if list(p.annots()):
  File "/home/ndntam/.local/lib/python3.10/site-packages/fitz/fitz.py", line 6701, in annots
    annot = self.load_annot(xref)
  File "/home/ndntam/.local/lib/python3.10/site-packages/fitz/fitz.py", line 6150, in load_annot
    val = self._load_annot(name, xref)
  File "/home/ndntam/.local/lib/python3.10/site-packages/fitz/fitz.py", line 6051, in _load_annot
    return _fitz.Page__load_annot(self, name, xref)
RuntimeError
```
If there was more further detail about the RuntimeError, I don't have it.  Its also possible that was an earlier version of PyMuPDF or MuPDF.

**Reply by JorjMcKie** [reply]
This is of course hard to tell without further context on version number and potentially example files.

My initial reaction to the chosen approach to answer the question itself ("page empty?") is, that it uses sledge hammers to crack nuts:
All the object types links, annots, widgets are forward chained. To determine whether there is at least one annot you shouldn't make a list of all annots, instead check `p.first_annot != None`. Same thing for link and widget.
Making a list of all annots entails multiple calls to the base library, which in most cases raises `RuntimeError` if the PDF has internal problems somewhere deep inside.
BTW I doubt that the above error log is complete, because if the base library raises exceptions, then there usually also is some message starting like `mupdf: ...`.

> The best performance with the least chance to raise unnecessary exceptions, is checking whether the page's `/Annots` is a non-empty PDF array. As I wrote earlier, all links, annots, and widgets are represented in that array with their xref numbers. This check will not create **_any_** of the object types `Link`, `Annot` or `Widget`, but actually look at the PDF's "source code":
```python
if p.annot_xrefs():
    msgs.append("Has annotations, links or widgets")
if p.get_text("text"):
        msgs.append("Has text")
```

**Reply by cbm755** [reply]
Thank you!  I have gone with [a minimal fix for now](https://gitlab.com/plom/plom/-/merge_requests/2084), but may try to the`annot_xref()` approach.

Replacing the "sledge hammer" is #815.

**Reply by JorjMcKie** [reply]
Maybe my formulation was a bit drastic ... 😉. What I was aiming at was that we need not try **_extracting_** annotation objects if all we want to know is whether the page **_has_** annotations at all.

And especially we need not extract **_all_** of them all for that check:
It suffices to check whether there exists the **_start_** of the annotation chain `doc[0].first_annot != None`. Similarly, there are `.first_link` and `.first_widget`.
These checks would of course still create one of the respective objects **_if_** they exist.
Checking the emptiness of list `doc[0].annot_xrefs()` in contrast will not create any of those and is therefore the fastest option.

Note:
Your post has actually ignited some internal discussions, whether we should / could do something about that "orphaning" mechanism behind `doc[0].annots()`. Along the line how the deletion of page `doc[0]` can be postponed, until the last annotation created by the iterator `doc[0].annots()` goes out of existence.
This is more complex than one would assume at first sight, because every `Document`, `Page`, `Annot`, `Link` and `Widget` Python object is shadowed by a corrresponding C object inside MuPDF. The `__del__()` Python methods remove the corresponding MuPDF object. These deletions must happen in a way that respects the intrinsic object hierarchy - or else the Python interpreter crashes as soon as an annotation deletion is being tried, for which the owning MuPDF page no longer exists.
This is currently ensured by that orphaning mechanism.
Not sure where all this will lead us to, but let's see.
